### PR TITLE
refactor JavaModel.getTarget()

### PR DESCRIPTION
**Pull request by jukzi**
to take heavy weight arguments.

As a preparation for performance fix:
For ClasspathEntry of type CPE_LIBRARY JavaModel.isExternalFile() is costly to compute and should be cached inside ClasspathEntry

https://github.com/eclipse-jdt/eclipse.jdt.core/issues/303

### CODE REVIEW

File: org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/JavaModel.java [hunk 1]
```diff
@@ -406,6 +413,7 @@ public static boolean isJmod(File file) {
  * JavaModelManager's external file cache.
  */
 static private boolean isExternalFile(IPath path) {
+	System.out.println("isExternalFile: "+path); //$NON-NLS-1$
```
Comment (iloveeclipse) [inline org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/JavaModel.java:416]: Please don't forget to remove for final version

File: org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/JavaModel.java [hunk 2]
```diff
@@ -337,6 +337,13 @@ public static Object getTarget(IPath path, boolean checkResourceExistence) {
 		return target;
 	return getExternalTarget(path, checkResourceExistence);
 }
+public static Object getTarget(IClasspathEntry entry, boolean checkResourceExistence) {
+	return getExternalTarget(entry.getPath(), checkResourceExistence);
```
Comment (iloveeclipse) [inline org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/JavaModel.java:341]: I believe this is not correct to skip `getWorkspaceTarget()` here and assume the classpath entry can't represent a path in the workspace.

Same below.

**Review by iloveeclipse — CHANGES_REQUESTED** [review]

File: org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/JavaModel.java (hunk 1 above)
Comment (jukzi) [reply org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/JavaModel.java:416]: oh, yes, thanks!

File: org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/JavaModel.java [hunk 3]
```diff
@@ -337,6 +337,13 @@ public static Object getTarget(IPath path, boolean checkResourceExistence) {
 		return target;
 	return getExternalTarget(path, checkResourceExistence);
 }
+public static Object getTarget(IClasspathEntry entry, boolean checkResourceExistence) {
```
[the remark this answers is not in the document]
Comment (jukzi) [reply org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/JavaModel.java:341]: i put links in

**Pull request merged**